integration-tests/hibernate-orm-panache-rest/src/main/java/io/quarkus/it/panache/rest/AuthorController.java

```diff
@@ -0,0 +1,18 @@
+package io.quarkus.it.panache.rest;
+
+import javax.ws.rs.core.Response;
+
+import io.quarkus.hibernate.orm.panache.rest.PanacheCrudController;
+import io.quarkus.hibernate.orm.panache.rest.PanacheRestResource;
+
+public interface AuthorController extends PanacheCrudController<Author, Long> {
+
+    @PanacheRestResource(exposed = false)
```

Comment:
If we would go away from an override, maybe it would be better to consider avoiding extending an interface all together? User could specify what methods to expose with the annotations like `@PanacheList`, `@PanacheAdd` etc.

Plus in the future we could add `@PanacheListBy(fieldName="")`.

I just wonder what is the more appropriate way for Quarkus? Annotations or methods?

Reply:
Well, IMO override is for overriding the behaviour of the method. Change its implementation. This is disabling, so rewriting the entire signature feels like overkill, no?

Reply:
I agree, that's why I'm raising the point of maybe getting rid of override in the first place and relying only on annotations to define which methods should be implemented. Since all the definitions are in the interfaces user cannot really override any implementation anyway (at lest not at the moment).

Reply:
Well I imagine you can override it by declaring a default method.

OTOH if you want to override it and at the same time require more parameters such as query params or context objects then you can't use an override and your approach with annotating a method with `@PanacheList` (which would automatically remove the `list` generation we would normally add) sounds indeed like a more flexible approach.

Reply:
@geoand want to pitch in?

Reply:
I like @FroMage's approach more TBH. I don't think we need to make things fully flexible (otherwise the user could just create their own Resource, no?) and  the `@Crud` seems to me like a good idea to add just enough flexibility

Reply:
>  This is disabling, so rewriting the entire signature feels like overkill
Also agree: to disable a method, some annotation on the interface level should be OK. (maybe `@Crud` is too generic, `@RestCrud` seems better as it's a CRUD over REST right?) 

But as I said in a comment, this is very common to add security, openapi and metrics to a REST endpoint, so a user may ends up overriding very often those methods (except if we provides `@EnablesMetrics`, `@EnableOpenApi` and `@EnabledSecurity` annotations that triggers generation of default implementation for them ...)

Reply:
So, it depends on how you would write code to override an otherwise-generate method. Which comes down to our discussion about creating an override or using `@PanacheList` to annotate a method replacement.
In either approaches, do we allow calling the original method (as declared in the `PanacheCrudController` interface) to get the default implementation (and thus, just change the annotations) or would that method throw?
I guess we do need to test it out for size.

Reply:
I would not allow to provides a custom implementation at this stage, but only to allow to override the method definition inside the interface to add security, open-api and metrics annotation.

Something like this should works
```
public interface AuthorController extends PanacheCrudController<Author, Long> {
  @Metered
  @RolesAllowed("READER")
  @Operation(summary = "List all authors")
  @Override
  Response list();
}
```

Reply:
Hm... the more use cases come up, the more it does seem that we have to allow users to add custom annotations on methods... So it seems like having something `@Crud` on the class won't fly well, since we would need to adding such annotations for things like OpenAPI, metrics etc...
So maybe just overriding the method is the way to go. What are the chances that a user will get the signature wrong when you have IDEs to help you (not to mention the `@Override` annotation that the compiler will checj)?

Reply:
The I'd suggest to keep `enable/disable` as is as well.

Reply:
Well, there are really three different use-cases here.

1/ Keeping the behaviour but altering annotations:

```java
public interface AuthorController extends PanacheCrudController<Author, Long> {
// TODO: are the original annotations inherited? @GET, etc…
  @Metered
  @RolesAllowed("READER")
  @Operation(summary = "List all authors")
  @Override
  Response list();
}
```

2/ Changing the behaviour by delegating to original impl:

```java
public interface AuthorController extends PanacheCrudController<Author, Long> {
// TODO: original annotations?
  @Metered
  @RolesAllowed("READER")
  @Operation(summary = "List all authors")
  @Override
  default Response list(){
    Logger.info("We're doing a list, baby, yeah!");
    return PanacheCrudController::super.list(); // ATM this won't fly since it's not default
  }
}
```

3/ Changing the behaviour entirely:

```java
public interface AuthorController extends PanacheCrudController<Author, Long> {
// TODO: original annotations?
  @Metered
  @RolesAllowed("READER")
  @Operation(summary = "List all authors")
  @SomeAnnotationToSayThisReplacesList
  default Response list(@QueryParam String user){
    Logger.info("We're doing a list for %s, baby, yeah!", user);
    return Response.ok(Author.byName(user).list()).build();
  }
}
```

All three use-cases have open questions:

1/ Do we have to repeat the original annotations? I guess yes otherwise we can't change/remove them
2/ How can we delegate to the original method? ATM we can't because the method is not defaulted, and if it's not we can't invoke it. Supporting the use-case from 3/ would allow us to specify the new method with a new name and still generate (and manually invoke) the original method (unexposed).
3/ We need an annotation to specify that this method replaces another one, but it allows us to extend the parameters and solve 2/ as well.

None of those requirements force us towards any direction as to how to disable one method (as opposed to altering it), and so I still think an annotation on the class is easier than specifying the signature just to remove it. It feels really awkward.

Reply:
`PanacheCrudController` does not have any annotations, they are added automatically when generating implementation. There were issues with generics declaration and RestEasy ended up registering two resources for the same path. For example with POST it would register one resource at `/books` that accepts `Book` as a body and another that accepts `Object`. If there is a way to work around this I'm happy to look into it again.

Also right now none of these `default` implementations will work. Code is always generated for these methods. Of course this is something that could be changed in the future.

Reply:
> PanacheCrudController does not have any annotations, they are added automatically when generating implementation

Yeah, I know. What I'm asking is should the enhancer add those annotations here too? If we don't add them the endpoint method will not work if you override it only with new annotations. OTOH if we do add them (at build-time) then the user can't remove them or alter them (replace a `PUT` by a `PATCH`).

> Also right now none of these default implementations will work. Code is always generated for these methods. Of course this is something that could be changed in the future.

Sure. And it's fine to add support for overriding them later, but if possible we could agree on a strategy right now and implement it later, this way users won't have to rewrite their code.